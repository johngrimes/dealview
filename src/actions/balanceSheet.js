// @flow

import { putObject, getObject } from 'db/db'
import type { BalanceSheetOverTime } from 'types/balanceSheet'
import type { AssetMap } from 'types/assets/asset'
import type { LiabilityMap } from 'types/liabilities/liability'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'BalanceSheet'
const theOneAndOnlyKey = 'BalanceSheet'

type CalculateBalanceSheetTask = {
  task: 'CALCULATE_BALANCE_SHEET',
  assets: AssetMap,
  liabilities: LiabilityMap,
  startDate: string,
  endDate: string,
}

type LoadBalanceSheetRequestAction = {
  type: 'LOAD_BALANCE_SHEET_REQUEST'
}
type LoadBalanceSheetSuccessAction = {
  type: 'LOAD_BALANCE_SHEET_SUCCESS',
  balanceSheet: BalanceSheetOverTime
}
type LoadBalanceSheetFailureAction = {
  type: 'LOAD_BALANCE_SHEET_FAILURE',
  error: string|null
}

type UpdateBalanceSheetRequestAction = {
  type: 'UPDATE_BALANCE_SHEET_REQUEST',
  assets: AssetMap,
  liabilities: LiabilityMap,
  startDate: string,
  endDate: string,
}
type UpdateBalanceSheetSuccessAction = {
  type: 'UPDATE_BALANCE_SHEET_SUCCESS',
  balanceSheet: BalanceSheetOverTime,
}
type UpdateBalanceSheetFailureAction = {
  type: 'UPDATE_BALANCE_SHEET_FAILURE',
  error: string|null,
}

type InvalidateBalanceSheetAction = {
  type: 'INVALIDATE_BALANCE_SHEET'
}

export type BalanceSheetAction = LoadBalanceSheetRequestAction
                               | LoadBalanceSheetSuccessAction
                               | LoadBalanceSheetFailureAction
                               | UpdateBalanceSheetRequestAction
                               | UpdateBalanceSheetSuccessAction
                               | UpdateBalanceSheetFailureAction
                               | InvalidateBalanceSheetAction

export const calculateBalanceSheet = (assets: AssetMap, liabilities: LiabilityMap, startDate: string, endDate: string): CalculateBalanceSheetTask => {
  return {
    task: 'CALCULATE_BALANCE_SHEET',
    assets,
    liabilities,
    startDate,
    endDate,
  }
}

export const loadBalanceSheetRequest = (): LoadBalanceSheetRequestAction => {
  return { type: 'LOAD_BALANCE_SHEET_REQUEST' }
}

export const loadBalanceSheetSuccess = (balanceSheet: BalanceSheetOverTime): LoadBalanceSheetSuccessAction => {
  return {
    type: 'LOAD_BALANCE_SHEET_SUCCESS',
    balanceSheet,
  }
}

export const loadBalanceSheetFailure = (error: string|null): LoadBalanceSheetFailureAction => {
  return {
    type: 'LOAD_BALANCE_SHEET_FAILURE',
    error,
  }
}

export const updateBalanceSheetRequest = (assets: AssetMap, liabilities: LiabilityMap, startDate: string, endDate: string): UpdateBalanceSheetRequestAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_REQUEST',
    assets,
    liabilities,
    startDate,
    endDate,
  }
}

export const updateBalanceSheetSuccess = (balanceSheet: BalanceSheetOverTime): UpdateBalanceSheetSuccessAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_SUCCESS',
    balanceSheet,
  }
}

export const updateBalanceSheetFailure = (error: string|null): UpdateBalanceSheetFailureAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_FAILURE',
    error,
  }
}

export const invalidateBalanceSheet = (): InvalidateBalanceSheetAction => {
  return { type: 'INVALIDATE_BALANCE_SHEET' }
}

export const updateBalanceSheet = (assets: AssetMap, liabilities: LiabilityMap,
                                   startDate: string, endDate: string): Thunk => {
  return dispatch => {
    dispatch(updateBalanceSheetRequest(assets, liabilities, startDate, endDate))
    return new Promise(async (resolve, reject) => {
      try {
        const worker = new Worker('/static/js/worker.js', { name: 'Balance Sheet Worker' })
        worker.onmessage = async event => {
          console.log('Balance sheet calc returned:', event.data)
          // flow-ignore
          const saved = await putObject(objectStore, event.data, theOneAndOnlyKey)
          dispatch(updateBalanceSheetSuccess(saved))
          resolve(saved)
        }
        worker.postMessage([ assets, liabilities, startDate, endDate ])
      } catch (error) {
        dispatch(updateBalanceSheetFailure(error))
        reject(error)
      }
    })
  }
}

export const loadBalanceSheet = (): Thunk => {
  return dispatch => {
    dispatch(loadBalanceSheetRequest())
    return new Promise(async (resolve, reject) => {
      try {
        const balanceSheet = await getObject(objectStore, theOneAndOnlyKey)
        dispatch(loadBalanceSheetSuccess(balanceSheet))
        resolve(balanceSheet)
      } catch (error) {
        dispatch(loadBalanceSheetFailure(error))
        reject(error)
      }
    })
  }
}
