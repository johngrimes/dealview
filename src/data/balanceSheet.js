import values from 'lodash.values'
import arrayHead from 'lodash.head'
import arrayTail from 'lodash.tail'
import moment from 'moment'

import { DateStorageFormat } from './commonTypes.js'
import { getValuationAtDate } from './asset.js'
import { getBalanceAtDate } from './loan.js'

export const balanceSheetOverTime = (
  assets,
  liabilities,
  startDate,
  endDate
) => {
  return Object.assign(
    {},
    ...calcBalanceSheet(assets, liabilities, startDate, endDate)
  )
}

function * calcBalanceSheet(assets, loans, date, endDate) {
  let nextDate = date
  while (!nextDate.isAfter(endDate)) {
    const totalAssets = sumAssetValueAtDate(values(assets), nextDate)
    const totalLiabilities = sumLoanBalanceAtDate(values(loans), nextDate)
    yield {
      [nextDate.format(DateStorageFormat)]: {
        totalAssets,
        totalLiabilities,
        equity: totalAssets - totalLiabilities,
      },
    }
    nextDate = moment(nextDate).add(1, 'days')
  }
}

const sumAssetValueAtDate = (assets, date) => {
  if (assets.length === 0) return 0
  const { head, tail } = { head: arrayHead(assets), tail: arrayTail(assets) }
  const headResult = getValuationAtDate(head, date)
  return headResult + sumAssetValueAtDate(tail, date)
}

const sumLoanBalanceAtDate = (loans, date) => {
  if (loans.length === 0) return 0
  const { head, tail } = { head: arrayHead(loans), tail: arrayTail(loans) }
  const headResult = getBalanceAtDate(head, date)
  return headResult + sumLoanBalanceAtDate(tail, date)
}
