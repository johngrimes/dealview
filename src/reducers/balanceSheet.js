// @flow

import type { Reducer, State, Action } from 'redux'

import type { BalanceSheetOverTime } from 'data/balanceSheet'
import type { BalanceSheetAction } from 'actions/balanceSheet'

export type BalanceSheetState = {
  status: 'uninitialised'|'loading'|'loaded'|'error',
  error?: string,
  balanceSheet: BalanceSheetOverTime
}
const initialState: BalanceSheetState = {
  status: 'uninitialised',
  balanceSheet: {},
}

const BalanceSheetReducer: Reducer<State, Action> = (state = initialState, action: BalanceSheetAction) => {
  switch (action.type) {
    case 'LOAD_BALANCE_SHEET_REQUEST':
    case 'UPDATE_BALANCE_SHEET_REQUEST':
      return {
        ...state,
        status: 'loading',
      }
    case 'LOAD_BALANCE_SHEET_SUCCESS':
    case 'UPDATE_BALANCE_SHEET_SUCCESS':
      return {
        ...state,
        status: 'loaded',
        balanceSheet: action.balanceSheet,
      }
    case 'LOAD_BALANCE_SHEET_FAILURE':
    case 'UPDATE_BALANCE_SHEET_FAILURE':
      return {
        ...state,
        status: 'error',
        error: action.error,
      }
    default:
      return state
  }
}

export default BalanceSheetReducer
