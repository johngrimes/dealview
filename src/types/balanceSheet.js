// @flow

import _ from 'lodash'
import moment from 'moment'
import type { Moment } from 'moment'

import { DateFormat } from 'types/commonTypes'
import { getValuationAtDate } from 'types/assets/asset'
import type { Asset, AssetMap } from 'types/assets/asset'
import type { LiabilityMap } from 'types/liabilities/liability'

export type BalanceSheet = {
  totalAssets: number,
  totalLiabilities: number,
  equity: number
}
export type BalanceSheetOverTime = { [date: string]: BalanceSheet }

export const balanceSheetOverTime = (assets: AssetMap, liabilities: LiabilityMap,
                                     startDate: string, endDate: string): BalanceSheetOverTime => {
  const startDay = moment(startDate, DateFormat)
  const endDay = moment(endDate, DateFormat)
  return calcBalanceSheet(assets, liabilities, startDay, endDay)
}

const calcBalanceSheet = (assets: AssetMap, liabilities: LiabilityMap,
                          date: Moment, endDate: Moment): BalanceSheetOverTime => {
  if (date.isAfter(endDate)) return {}
  const totalAssets = sumAssetValueAtDate(_.values(assets), date)
  const totalLiabilities = 0  // sumLiabilityValueAtDate(liabilities, date)
  const nextDate = moment(date).add(1, 'days')
  return { ...calcBalanceSheet(assets, liabilities, nextDate, endDate),
    [date.format(DateFormat)]: {
      totalAssets,
      totalLiabilities,
      equity: (totalAssets - totalLiabilities)
    }
  }
}

const sumAssetValueAtDate = (assets: Asset[], date: string): number => {
  if (assets.length === 0) return 0
  const { head, tail } = { head: _.head(assets), tail: _.tail(assets) }
  return getValuationAtDate(head, date) + sumAssetValueAtDate(tail, date)
}
