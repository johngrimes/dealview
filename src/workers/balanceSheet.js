// @flow

import moment from 'moment'

import { balanceSheetOverTime } from 'types/balanceSheet'
import { DateStorageFormat } from 'types/commonTypes'

onmessage = event => {
  console.log(`Balance sheet calc started:`, event.data)
  // flow-ignore
  postMessage(balanceSheetOverTime(
    event.data[0],
    event.data[1],
    moment(event.data[2], DateStorageFormat),
    moment(event.data[3], DateStorageFormat),
  ))
}
