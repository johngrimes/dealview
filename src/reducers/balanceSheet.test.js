import BalanceSheetReducer from './balanceSheet.js'
import * as BalanceSheetActions from '../actions/balanceSheet.js'
import {
  validBalanceSheet1,
  validBalanceSheet2,
} from '../data/fixtures/balanceSheet.js'

describe('BalanceSheetReducer', () => {
  const initialState = {
    status: 'loaded',
    fresh: true,
    balanceSheet: validBalanceSheet1,
  }

  it('should update state upon request', () => {
    const actions = [
      BalanceSheetActions.updateBalanceSheetRequest,
      BalanceSheetActions.loadBalanceSheetRequest,
    ]
    actions.forEach(action => {
      const nextState = BalanceSheetReducer(initialState, action())
      expect(nextState).toMatchSnapshot()
    })
  })

  it('should update state upon update success', () => {
    const action = BalanceSheetActions.updateBalanceSheetSuccess(
      validBalanceSheet2
    )
    const nextState = BalanceSheetReducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })

  it('should freshen upon update success', () => {
    const action = BalanceSheetActions.updateBalanceSheetSuccess(
      validBalanceSheet2
    )
    const nextState = BalanceSheetReducer(
      { ...initialState, fresh: false },
      action
    )
    expect(nextState).toMatchSnapshot()
  })

  it('should update state upon load success', () => {
    const action = BalanceSheetActions.loadBalanceSheetSuccess(
      validBalanceSheet2
    )
    const nextState = BalanceSheetReducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })

  it('should update state upon failure', () => {
    const actions = [
      BalanceSheetActions.updateBalanceSheetFailure,
      BalanceSheetActions.loadBalanceSheetFailure,
    ]
    actions.forEach(action => {
      const nextState = BalanceSheetReducer(initialState, action('Some error'))
      expect(nextState).toMatchSnapshot()
    })
  })

  it('should invalidate balance sheet', () => {
    const action = BalanceSheetActions.invalidateBalanceSheet()
    const nextState = BalanceSheetReducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })
})
