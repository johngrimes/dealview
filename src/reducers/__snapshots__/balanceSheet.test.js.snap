// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalanceSheetReducer should update state upon failure 1`] = `
Object {
  "balanceSheet": Object {
    "2017-01-01": Object {
      "equity": 155070,
      "totalAssets": 1000070,
      "totalLiabilities": 845000,
    },
    "2017-01-02": Object {
      "equity": 165080,
      "totalAssets": 1000080,
      "totalLiabilities": 835000,
    },
  },
  "error": "Some error",
  "status": "error",
}
`;

exports[`BalanceSheetReducer should update state upon failure 2`] = `
Object {
  "balanceSheet": Object {
    "2017-01-01": Object {
      "equity": 155070,
      "totalAssets": 1000070,
      "totalLiabilities": 845000,
    },
    "2017-01-02": Object {
      "equity": 165080,
      "totalAssets": 1000080,
      "totalLiabilities": 835000,
    },
  },
  "error": "Some error",
  "status": "error",
}
`;

exports[`BalanceSheetReducer should update state upon load success 1`] = `
Object {
  "balanceSheet": Object {
    "2017-01-01": Object {
      "equity": 155070,
      "totalAssets": 1000070,
      "totalLiabilities": 845000,
    },
    "2017-01-02": Object {
      "equity": 165080,
      "totalAssets": 1000080,
      "totalLiabilities": 835000,
    },
    "2017-01-03": Object {
      "equity": 165080,
      "totalAssets": 1000080,
      "totalLiabilities": 835000,
    },
  },
  "status": "loaded",
}
`;

exports[`BalanceSheetReducer should update state upon request 1`] = `
Object {
  "balanceSheet": Object {
    "2017-01-01": Object {
      "equity": 155070,
      "totalAssets": 1000070,
      "totalLiabilities": 845000,
    },
    "2017-01-02": Object {
      "equity": 165080,
      "totalAssets": 1000080,
      "totalLiabilities": 835000,
    },
  },
  "status": "loading",
}
`;

exports[`BalanceSheetReducer should update state upon request 2`] = `
Object {
  "balanceSheet": Object {
    "2017-01-01": Object {
      "equity": 155070,
      "totalAssets": 1000070,
      "totalLiabilities": 845000,
    },
    "2017-01-02": Object {
      "equity": 165080,
      "totalAssets": 1000080,
      "totalLiabilities": 835000,
    },
  },
  "status": "loading",
}
`;

exports[`BalanceSheetReducer should update state upon update success 1`] = `
Object {
  "balanceSheet": Object {
    "2017-01-01": Object {
      "equity": 155070,
      "totalAssets": 1000070,
      "totalLiabilities": 845000,
    },
    "2017-01-02": Object {
      "equity": 165080,
      "totalAssets": 1000080,
      "totalLiabilities": 835000,
    },
    "2017-01-03": Object {
      "equity": 165080,
      "totalAssets": 1000080,
      "totalLiabilities": 835000,
    },
  },
  "status": "loaded",
}
`;
